sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -589,22 +593,37 @@ case class DateFormatClass(left: Expression, right: Expression, timeZoneId: Opti
 
   override def inputTypes: Seq[AbstractDataType] = Seq(TimestampType, StringType)
 
-  override def withTimeZone(timeZoneId: String): TimeZoneAwareExpression =
-    copy(timeZoneId = Option(timeZoneId))
+  override def withTimeZone(timeZoneId: String): TimeZoneAwareExpression = {
+    val tf = if (formatter.isEmpty && right.foldable) {
+      val format = right.eval().toString
```

Comment:
@MaxGekk, can `right.eval()` be `null`? It's pretty much a corner case, for instance, `cast(null as string)` but might better handle it.